drivers/video/ov5640.c

```diff
@@ -457,23 +635,84 @@ static int ov5640_write_multi_regs(const struct i2c_dt_spec *spec, const struct
 	return 0;
 }
 
+static int ov5640_set_fmt_dvp(const struct ov5640_config *cfg)
+{
+	uint8_t reg;
+	int ret = 0;
+
+	ret |= ov5640_read_reg(&cfg->i2c, TIMING_TC_REG21_REG, &reg, sizeof(reg));
+	ret |= ov5640_write_reg(&cfg->i2c, TIMING_TC_REG21_REG, (reg & 0xDF) | 0x00);
+
+	if (ret) {
+		LOG_ERR("Unable to configure REG: %d on DVP", TIMING_TC_REG21_REG);
+		return ret;
+	}
+
+	ret |= ov5640_read_reg(&cfg->i2c, SYS_RESET02_REG, &reg, sizeof(reg));
+	ret |= ov5640_write_reg(&cfg->i2c, SYS_RESET02_REG, (reg & 0xE3) | 0x1C);
+
+	if (ret) {
+		LOG_ERR("Unable to configure REG: %d on DVP", SYS_RESET02_REG);
+		return ret;
+	}
+
+	ret |= ov5640_read_reg(&cfg->i2c, SYS_CLK_ENABLE02_REG, &reg, sizeof(reg));
+	ret |= ov5640_write_reg(&cfg->i2c, SYS_CLK_ENABLE02_REG, (reg & 0xD7) | 0x00);
+
+	if (ret) {
+		LOG_ERR("Unable to configure REG: %d on DVP", SYS_CLK_ENABLE02_REG);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int ov5640_set_fmt_csi2(const struct ov5640_config *cfg,
+				const struct ov5640_resolution_config *resolution_params)
+{
+	int ret = 0;
+
+	/* Configure MIPI pixel clock */
+	ret |= ov5640_modify_reg(&cfg->i2c, SC_PLL_CTRL0_REG, 0x0f, 0x08);
+	ret |= ov5640_modify_reg(&cfg->i2c, SC_PLL_CTRL1_REG, 0xff,
+				 resolution_params->mipi_pclk.pllCtrl1);
+	ret |= ov5640_modify_reg(&cfg->i2c, SC_PLL_CTRL2_REG, 0xff,
+				 resolution_params->mipi_pclk.pllCtrl2);
+	ret |= ov5640_modify_reg(&cfg->i2c, SC_PLL_CTRL3_REG, 0x1f, 0x13);
+	ret |= ov5640_modify_reg(&cfg->i2c, SYS_ROOT_DIV_REG, 0x3f, 0x01);
+	ret |= ov5640_write_reg(&cfg->i2c, PCLK_PERIOD_REG, 0x0a);
+
+	if (ret) {
+		LOG_ERR("Unable to configure MIPI pixel clock");
+		return ret;
+	}
+
+	return 0;
+}
+
 static int ov5640_set_fmt(const struct device *dev, enum video_endpoint_id ep,
 			  struct video_format *fmt)
 {
 	struct ov5640_data *drv_data = dev->data;
 	const struct ov5640_config *cfg = dev->config;
+	const struct video_format_cap *fmts = ov5640_is_dvp(dev) ? dvp_fmts : csi2_fmts;
+	const size_t num_fmts = ov5640_is_dvp(dev) ? ARRAY_SIZE(dvp_fmts) : ARRAY_SIZE(csi2_fmts);
+	const struct ov5640_resolution_config *resolution_params = ov5640_is_dvp(dev) ?
+						dvp_resolution_params : csi2_resolution_params;
+	const size_t num_res_params = ov5640_is_dvp(dev) ?
+			ARRAY_SIZE(dvp_resolution_params) : ARRAY_SIZE(csi2_resolution_params);
```

Comment:
Minor style comment: I am tempted by a single `if/else` or `switch (cfg->bus_type)` to help with readability and error handling.

Reply:
Yeap. It's better. I'll change to `if/else`.